### PR TITLE
Reworked NetCode functions, updated folder structure.

### PR DESCRIPTION
**Pull request by QUB3X**
- Added a JSON configuration for OpCodes, ```opcodes.json```.

- Moved file of classes inside ```classes/```

- Changed implementation of NetCode.js with new function ```NetCode.Do()```, which handles sending messages to clients in a more elegant way

- Soon to be added unit tests

### CODE REVIEW

File: classes/NetCode.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            // Do a broadcast
+            for (let player of target) {
+                if (player.uid !== except) {
+                    sender.send(buffer, player.socket.port, player.socket.address)
+                }
+            }
+        } else {
+            // If it's a single target
+            if( typeof target.socket.port === "undefined" &&
+                typeof target.socket.address === "undefined") {
```
Comment (YoDevil) [inline classes/NetCode.js]: I'm pretty sure this throws an exception: trying to access .port of undefined.
You should just check target.socket, or, better, check with instanceof

File: classes/NetCode.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+for (let id in opcodes) {
+    OpCode[id] = opcodes[id].opcode
+}
+
+// Freeze the reference object, thus making it a enum
+Object.freeze(OpCode)
+
+
+console.log("Listing all defined opcodes:")
+console.log(OpCode)
```
Comment (YoDevil) [inline classes/NetCode.js]: lets not

File: classes/NetCode.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+
+class NetCode {
+    /**
+     * Send a command to the client(s).
+     * @param {string} opcode The name of the operation you want to perform.
+     * @param {[any]} params Array of parameters.
+     * @param {this.server} sender The sender object (usually the server socket)
+     * @param {Player | Socket} target The destination of the command (can be a player of a group of players, or a socket)
+     * @param {*} except An entity present in target which should be excluded
```
Comment (YoDevil) [inline classes/NetCode.js]: It's a Player, as when using it the previous parameter is an array of players

File: classes/NetCode.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        return Object.keys(OpCode)[opcode]
+    }
+
+    /**
+     * Write the port of the lobby in the buffer,
+     * after finding an online lobby.
+     * @param {*} socket The server socket
+     * @param {*} lobby The port of the lobby that gets online
+     */
+    WriteBufferOnSocket(socket, lobby) {
```
Comment (YoDevil) [inline classes/NetCode.js]: What

File: index.js [hunk 5]
```diff
@@ -1,89 +1,91 @@
...
 
 const server = net.createServer(socket => {
-    console.log("New client connected");
-    socket.on("error", err => console.log(err));
+    console.log("New client connected")
+
+
+    socket.on("error", err => console.log(err))
+
+
```
Comment (YoDevil) [inline index.js]: remove

File: index.js [hunk 6]
```diff
@@ -1,89 +1,91 @@
...
 
-const lobbies = {}; // Object of lobbies, port: process
+const lobbies = {} // Object of lobbies, port: process
 
 const server = net.createServer(socket => {
-    console.log("New client connected");
-    socket.on("error", err => console.log(err));
+    console.log("New client connected")
+
+
```
Comment (YoDevil) [inline index.js]: remove

File: index.js [hunk 7]
```diff
@@ -1,89 +1,91 @@
...
-                    if(!lobbies[lobby].online) return;
-                    let bufferView = NetCode.BufferOp(OpCode.FoundMatch, 4);
-                    bufferView.setUint16(1, lobby, true);
-                    socket.write(Buffer.from(bufferView.buffer));
+                let watch = setInterval(() => {
+                    if (!lobbies[lobby].online) return
 
-                    clearInterval(watch);
-                }, 20);
+                    WriteBufferOnSocket(socket, lobby)
```
Comment (YoDevil) [inline index.js]: What

File: classes/NetCode.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            }
+        }
+    }
+
+    /**
+     * 
+     * @param {Int} opcode The OpCode.X enum value
+     * @returns {String} The OpCode as string (X)
+     */
+    static ResolveOpCode(opcode) {
```
Comment (YoDevil) [inline classes/NetCode.js]: What uses this?
Also it does not resolve the op *code*, but rather the op *name*

File: classes/NetCode.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+     */
+    static Do(opcode, params, sender, target, except = null) {
+        let op = opcodes[opcode]
+
+        let buffer = new ArrayBuffer(1 + op.size * params.length / op.params.length)
+        let view = new DataView(buffer)
+
+        view.setUint8(0, op.opcode)
+
+        let padding = 1
```
Comment (YoDevil) [inline classes/NetCode.js]: This is strictly related to the line above, what's the point of the whitespace

File: classes/Game.js [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+     * @param {float} jumpHeight The height of the jump
+     */
+    Jump(uid, jumpHeight) {
+        for (let player of this.players) {
+            if (player.uid == uid) {
+                if (player.state.isGrounded)
+                    player.input.nextJump = jumpHeight
+            }
+        }
+
```
Comment (YoDevil) [inline classes/Game.js]: remove

File: classes/Game.js [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+                player.state.pos.x,
+                player.state.pos.y
+            ],
+            this.server, this.players)
+
+        // Tell the new spawned player's client to spawn all the others players
+        for (let other of this.players) {
+            if (other.uid != player.uid) {
+                NC.Do("Spawn",
+                    [other.uid,
```
Comment (YoDevil) [inline classes/Game.js]: ugly

File: classes/NetCode.js [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+class NetCode {
+    /**
+     * Send a command to the client(s).
+     * @param {string} opcode The name of the operation you want to perform.
+     * @param {[any]} params Array of parameters.
+     * @param {this.server} sender The sender object (usually the server socket)
+     * @param {Player | Socket} target The destination of the command (can be a player of a group of players, or a socket)
+     * @param {*} except An entity present in target which should be excluded
+     */
+    static Do(opcode, params, sender, target, except = null) {
```
Comment (YoDevil) [inline classes/NetCode.js]: Can't we use a better name

File: index.js [hunk 13]
```diff
@@ -1,89 +1,91 @@
...
-const NetCode = require('./NetCode.js').NetCode;
-
-const MM_PORT = 50999;
-const BASE_PORT = 51000;
-const MAX_LOBBIES = 100;
-const MAX_PLAYER_COUNT = 2;
+const net = require('net')
+const fork = require('child_process').fork
+const OpCode = require('./classes/NetCode.js').OpCode
+const NC = require('./classes/NetCode.js').NetCode
```
Comment (YoDevil) [inline index.js]: Can we keep calling it NetCode please

File: classes/Game.js [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+            immediate: true
+        })
+
+        // Check for players who disconnects
+        let interval = 10 * 1000
+        setInterval(() => {
+            let time = Date.now()
+            for (let i = 0; i < this.players.length; i++) {
+                if (time - this.players[i].lastseen >= interval) {
+                    NC.Do("Disconnect", [
```
Comment (YoDevil) [inline classes/Game.js]: NC is undefined
Can we still call it NetCode please

File: classes/Game.js [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+
+    /**
+     * Disconnects a player from the server, removing it from the pool
+     * of players in that instance.
+     * @param {string} uid The UID of the player you want to disconnect.
+     */
+    Disconnect(uid) {
+        for (let i = 0; i < this.players.length; i++) {
+            if (this.players[i].uid == uid) {               
+                NC.Do("Disconnect", [
```
Comment (YoDevil) [inline classes/Game.js]: ditto

File: classes/Game.js [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+            x: Math.random() * 10 - 5,
+            y: 1
+        }
+        player.state.vel = {
+            x: 0,
+            y: 0
+        }
+
+        // Spawn the new player
+        NC.Do("Spawn", [
```
Comment (YoDevil) [inline classes/Game.js]: ditto

File: package.json [hunk 17]
```diff
@@ -5,7 +5,7 @@
   "main": "index.js",
   "scripts": {
     "start": "node index.js",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "jest"
```
Comment (YoDevil) [inline package.json]: Why is this in this PR

File: classes/NetCode.js [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+                case "UInt16":
+                    view.setUint16(padding, value, true)
+                    break
+                case "Float32":
+                default:
+                    view.setFloat32(padding, value, true)
+                    break
+            }
+
+            padding += op.size
```
Comment (YoDevil) [inline classes/NetCode.js]: What

File: lobby.js [hunk 19]
```diff
@@ -1,56 +1,61 @@
-const dgram = require('dgram');
-const OpCode = require('./NetCode.js').OpCode;
-const NetCode = require('./NetCode.js').NetCode;
-const Game = require('./Game.js');
+const dgram = require('dgram')
+const OpCode = require('./classes/NetCode.js').OpCode
+const NC = require('./classes/NetCode.js').NetCode
```
Comment (YoDevil) [inline lobby.js]: ditto

File: classes/NetCode.js [hunk 20]
```diff
@@ -0,0 +1,113 @@
+/**
+ * NetCode.js
+ */
+
+// Import JSON declaration of opcodes + types
+const opcodes = require("./../opcodes.json")
```
Comment (YoDevil) [inline classes/NetCode.js]: These are not opcodes though. I'd say they're "operations", "ops", "commands", "cmd", whatever
And the same goes for the filename

File: classes/Game.js [hunk 21]
```diff
@@ -0,0 +1,272 @@
+const Player = require('./Player.js')
+const NetCode = require('./../NetCode.js')
```
Comment (YoDevil) [inline classes/Game.js]: Wrong path

File: opcodes.json [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+        "opcode": 1,
+        "name": "Queue",
+        "params": [],
+        "size": 0
+    },
+    "FoundMatch": {
+        "opcode": 2,
+        "name": "FoundMatch",
+        "params": ["Uint16"],
+        "size": 2
```
Comment (YoDevil) [inline opcodes.json]: Doesn't size track the arguments *after* the first one which is the opcode?

File: opcodes.json [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+        "opcode": 3,
+        "name": "Register",
+        "params": [],
+        "size": 0
+    },
+    "Spawn": {
+        "opcode": 4,
+        "name": "Spawn",
+        "params": ["Uint16", "Float32", "Float32"],
+        "size": 10
```
Comment (YoDevil) [inline opcodes.json]: ditto

File: opcodes.json [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+        "opcode": 6,
+        "name": "Jump",
+        "params": [],
+        "size": 0
+    },
+    "SetPos": {
+        "opcode": 7,
+        "name": "SetPos",
+        "params": ["Uint16", "Float32", "Float32", "Float32", "Float32"],
+        "size": 18
```
Comment (YoDevil) [inline opcodes.json]: ditto

File: opcodes.json [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+    },
+    "Disconnect": {
+        "opcode": 8,
+        "name": "Disconnect",
+        "params": ["Uint16"],
+        "size": 2
```
Comment (YoDevil) [inline opcodes.json]: ditto

File: opcodes.json [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+    },
+    "Ping": {
+        "opcode": 9,
+        "name": "Ping",
+        "params": ["Int16"],
+        "size": 2
```
Comment (YoDevil) [inline opcodes.json]: ditto

**Comment by YoDevil** [comment]
Non usare --amend, fai pure fix commits che ho capito che github non tiene bene traccia delle discussioni aperte dopo un --amend

File: opcodes.json (hunk 22 above)
Comment (YoDevil) [reply opcodes.json]: And *params* should also only contain the parameters after the opcode

File: classes/NetCode.js (hunk 18 above)
Comment (YoDevil) [reply classes/NetCode.js]: Still wrong

File: package-lock.json [hunk 27]
```diff
@@ -1,23 +1,5505 @@
 {
   "name": "super-cool-server",
-  "version": "0.0.1",
+  "version": "0.0.2",
   "lockfileVersion": 1,
   "requires": true,
   "dependencies": {
+    "@babel/code-frame": {
```
Comment (YoDevil) [inline package-lock.json]: this whole thing has to go

File: index.js (hunk 6 above)
Comment (YoDevil) [reply index.js]: yay instead of less white we have more white now

File: classes/NetCode.js [hunk 28]
```diff
@@ -0,0 +1,107 @@
...
+                case "UInt16":
+                    view.setUint16(padding, value, true)
+                    break
+                case "Float32":
+                default:
+                    view.setFloat32(padding, value, true)
+                    break
+            }
+
+            padding += command.size
```
Comment (YoDevil) [inline classes/NetCode.js]: still wrong

File: classes/Game.js [hunk 29]
```diff
@@ -0,0 +1,268 @@
...
+                    ], this.server, this.players)
+
+                    this.players.splice(i, 1)
+                }
+            }
+        }, interval)
+    }
+
+    /**
+     * Connect a player to a socket and add it to the pool
```
Comment (YoDevil) [inline classes/Game.js]: "Create a player instance linked to its socket and..."

File: classes/Game.js [hunk 30]
```diff
@@ -0,0 +1,268 @@
...
+    Connect(socket) {
+        const player = new Player(socket)
+        this.players.push(player)
+        this.SendPlayerCount()
+    }
+
+    /**
+     * Disconnects a player from the server, removing it from the pool
+     * of players in that instance.
+     * @param {string} uid The UID of the player you want to disconnect.
```
Comment (YoDevil) [inline classes/Game.js]: definitely not a string

File: classes/Game.js [hunk 31]
```diff
@@ -0,0 +1,268 @@
...
+        process.send({
+            msg: "PlayerCount",
+            args: [this.players.length]
+        })
+    }
+
+    /**
+     * Refresh the .lastseen property of the player Object, which tracks the
+     * last time the player was seen online.
+     * @param {string} uid The UID of the player you want to update.
```
Comment (YoDevil) [inline classes/Game.js]: ditto

File: classes/Game.js [hunk 32]
```diff
@@ -0,0 +1,268 @@
...
+                player.lastseen = Date.now()
+            }
+        }
+    }
+
+    /**
+     * Tells the clients to spawn the new selected player, and
+     * then to spawn in the new player's client all the other
+     * players.
+     * @param {string} uid The UID of the player you want to spawn-
```
Comment (YoDevil) [inline classes/Game.js]: ditto

File: classes/Game.js [hunk 33]
```diff
@@ -0,0 +1,268 @@
...
+                if (player.state.isGrounded)
+                    player.input.nextJump = jumpHeight
+            }
+        }
+
+    }
+
+    /**
+     * Make the selected player move at a certain speed.
+     * @param {string} uid The UID of the player you want to move
```
Comment (YoDevil) [inline classes/Game.js]: definitely not a string

File: classes/NetCode.js [hunk 34]
```diff
@@ -0,0 +1,107 @@
...
+                sender.send(buffer, target.port, target.address)
+            } else {
+                // or is it a player?
+                sender.send(buffer, target.socket.port, target.socket.address)
+            }
+        }
+    }
+
+    /**
+     *
```
Comment (YoDevil) [inline classes/NetCode.js]: I guess you wanted to write something here

File: lobby.js [hunk 35]
```diff
@@ -1,56 +1,61 @@
...
+    // Receive a command from client
+    let receivedCmd = new DataView(data.buffer)
+    // If the command is empty, return
+    if (receivedCmd.byteLength <= 0) return
 
-    let op = receivedCmd.getInt8(0);
-    console.log(`got ${NetCode.OpCodeToString(op)} from ${client.address}:${client.port}`);
+    // 
+    let op = receivedCmd.getInt8(0)
+    console.log(`got ${NetCode.Resolve(op)} from ${client.address}:${client.port}`)
```
Comment (YoDevil) [inline lobby.js]: You named the function "ResolveOpCode"

File: classes/Game.js [hunk 36]
```diff
@@ -0,0 +1,268 @@
...
+        })
+
+        // Check for players who disconnects
+        let interval = 10 * 1000
+        setInterval(() => {
+            let time = Date.now()
+            for (let i = 0; i < this.players.length; i++) {
+                if (time - this.players[i].lastseen >= interval) {
+                    NetCode.Do("Disconnect", [
+                        this.players[i].uid
```
Comment (YoDevil) [inline classes/Game.js]: I think this would look better inline. You've done single parameters inline elsewhere, anyways.

File: classes/Game.js [hunk 37]
```diff
@@ -0,0 +1,264 @@
...
+            this.server, this.players)
+
+        // Tell the new spawned player's client to spawn all the others players
+        for (let other of this.players) {
+            if (other.uid != player.uid) {
+                NetCode.Do("Spawn", [
+                    other.uid,
+                    other.state.pos.x,
+                    other.state.pos.y],
+                    this.server, player)
```
Comment (YoDevil) [inline classes/Game.js]: just copy the above's syntax

File: lobby.js [hunk 38]
```diff
@@ -1,56 +1,61 @@
...
-    let receivedCmd = new DataView(data.buffer);
-    if(receivedCmd.byteLength <= 0) return;
+    // Receive a command from client
+    let receivedCmd = new DataView(data.buffer)
+    // If the command is empty, return
+    if (receivedCmd.byteLength <= 0) return
 
-    let op = receivedCmd.getInt8(0);
-    console.log(`got ${NetCode.OpCodeToString(op)} from ${client.address}:${client.port}`);
+    //
```
Comment (YoDevil) [inline lobby.js]: ?

File: package.json [hunk 39]
```diff
@@ -1,11 +1,10 @@
 {
   "name": "super-cool-server",
-  "version": "0.0.2",
+  "version": "0.1.0",
```
Comment (YoDevil) [inline package.json]: Doesn't match package-lock

File: classes/NetCode.js [hunk 40]
```diff
@@ -0,0 +1,91 @@
+/**
+ * NetCode.js
+ */
+
+// Import JSON declaration of commands + types
+const commands = require("./../commands.json")
+
+
+var OpCode = {}
```
Comment (YoDevil) [inline classes/NetCode.js]: let

File: commands.json [hunk 41]
```diff
@@ -0,0 +1,62 @@
+{
+    "Error": {
+        "opcode": 0,
+        "name": "Error",
```
Comment (YoDevil) [inline commands.json]: does anything use the "name" field?

Comment (QUB3X) [reply commands.json]: Actually no, but could be useful in the future or for debugging purposes :stuck_out_tongue_closed_eyes:

Comment (YoDevil) [reply commands.json]: It's redundant to have it both here and in the key

**Pull request closed without merging**